### PR TITLE
[Megatron-FSDP] Build default FSDP DeviceMesh, and remove model arg from fully_shard_optimizer().

### PR DESCRIPTION
**Pull request by cspades**
# What does this PR do ?

- Make all `DeviceMesh`-related arguments optional. If a DeviceMesh is not provided, an FSDP environment will be built for you by Megatron-FSDP.
- Remove the required (and irritating) `model` argument from `fully_shard_optimizer`, in favor of using a reference to `MegatronFSDP` via the optimizer's registered parameters: `optim_param._megatron_fsdp_model`
  - Backwards-incompatible change but a good change. Added some temporary `@experimental_api` decorators to bypass the CI/CD test, to be removed in a subsequent PR.

# Testing

- Works on BioNeMo ESM-2. Abbreviated loss curve.
```
wandb:                            train/loss █▅▄▃▄▄▂▃▄▃▃▃▃▂▂▃▃▃▂▁▂▁▂▂▃▃▃▃▂▂▃▂▂▂▁▁▃▂▂▂
```

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 1]
```diff
@@ -94,6 +95,26 @@ def fully_shard_model(
...
+        logger.warning(
+            "Building a generic FSDP DeviceMesh for Megatron-FSDP because "
+            f"DeviceMesh was not provided (device_mesh={device_mesh}). "
+        )
+        if dp_shard_dim is None:
+            dp_shard_dim = "fsdp"
+        # Deactivate DP-Outer, which needs to be consistent with Expert DeviceMesh.
+        dp_outer_dim = None
+        hybrid_fsdp_group = None
+        outer_dp_sharding_strategy = 0
```
Comment (cspades) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: This should turn off DP-Outer for everything in `FSDPDistributedIndex`, including the Expert DeviceMesh's DP-Outer dimension if it is provided.

But I have not tried passing an HSDP Expert DeviceMesh while also not passing a non-Expert DeviceMesh. Need to test it on a model with expert parameters.

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: Yes, the default device mesh should not include outer_dp_sharding_strategy.

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: Would it be clearer to write `outer_dp_sharding_strategy = ShardingStrategy.NO_SHARD` here?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 2]
```diff
@@ -94,6 +95,26 @@ def fully_shard_model(
     Returns:
         model (MegatronFSDP): The wrapped Megatron-FSDP model configured for FSDP.
     """
+    # If no DeviceMesh or FSDP dimension is provided, then build an FSDP DeviceMesh.
+    # Modify arguments into arguments necessary for vanilla FSDP.
+    if device_mesh is None:
+        logger.warning(
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: We don't need to notify the user that we are creating a device mesh.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 3]
```diff
@@ -94,6 +95,26 @@ def fully_shard_model(
...
+        # Deactivate DP-Outer, which needs to be consistent with Expert DeviceMesh.
+        dp_outer_dim = None
+        hybrid_fsdp_group = None
+        outer_dp_sharding_strategy = 0
+        device_mesh = init_device_mesh(
+            device_type="cuda",
+            mesh_shape=(torch.distributed.get_world_size(),),
+            mesh_dim_names=(dp_shard_dim,),
+        )
+        logger.info(f"Constructed the Megatron-FSDP DeviceMesh: {device_mesh}")
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: As above, this is the default behavior and does not need prompt here.

File: tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py [hunk 4]
```diff
@@ -591,3 +596,41 @@ def test_dcp_checkpoint_save_and_load(
...
+        )
+
+        # Construct toy model.
+        toy_model, toy_adam, fsdp_unit_modules = build_toy_model_and_optimizer(TRANSFORMER, False)
+
+        # Fully-shard the model and optimizer.
+        model = fully_shard_model(
+            module=toy_model, fsdp_unit_modules=fsdp_unit_modules, zero_dp_strategy=shard_strategy
+        )
+        optimizer = fully_shard_optimizer(model=model, optimizer=toy_adam)
```
Comment (shjwudp) [inline tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py]: Is it possible to make `fully_shard_optimizer` independent of the model? 

e.g.
```python
register_fsdp_optimizer_hook(optimizer=toy_adam)
```

Comment (cspades) [reply tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py]: I can look into it in this PR, will share what I come up with.

Comment (cspades) [reply tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py]: @shjwudp Yeah, I don't think we should do the optimizer update in this PR, most ideas I have involve many code changes. Let's merge this first.

Comment (shjwudp) [reply tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py]: If you prefer a mini-step approach, I think we can merge it first, but we should have such an API in place.

**Review by shjwudp — APPROVED** [review]

File: tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py (hunk 4 above)
Comment (cspades) [reply tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py]: Since we aligned on my idea, I'll do it in this PR!

Comment (cspades) [reply tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py]: Implemented! All unit tests pass on my side. Review time!
```
[default0]:==== 316 passed, 162 skipped, 6 xfailed, 430 warnings in 704.82s (0:11:44) =====
```

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 5]
```diff
@@ -227,45 +281,59 @@ def preprocess_dcp_and_te_extra_state(state_dict):
...
+    first_mfsdp_param = optimizer.param_groups[0][next(iter(optimizer.param_groups[0]))][0]
+    if not getattr(first_mfsdp_param, "_megatron_fsdp_model", None):
+        raise ValueError(
+            f"[MegatronFSDP fully_shard_optimizer()] Could not retrieve a reference to "
+            f"MegatronFSDP from the first registered Parameter: {first_mfsdp_param} \n"
+            "Make sure the optimizer is registered to the MegatronFSDP distributed "
+            "parameters via MegatronFSDP._replace_param_with_distributed_if_needed() "
+            "before initializing the optimizer on the MegatronFSDP model. "
+        )
+    mfsdp_api = first_mfsdp_param._megatron_fsdp_model
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: Why not just call it a `model` or `mfsdp_model`?

mfsdp_api appears to be a collection of functions rather than a model object.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 6]
```diff
@@ -533,8 +606,13 @@ def fully_shard(
...
 
     # Extend optimizer methods to support Megatron-FSDP operations.
-    if optimizer is not None:
-        fully_shard_optimizer(model, optimizer, preproc_state_dict_for_dcp_ckpt)
+    # Replace the optimizer module parameter references with
+    # Megatron-FSDP-managed distributed parameters.
+    model._replace_param_with_distributed_if_needed()
+    optimizer.param_groups.clear()
+    optimizer.state.clear()
+    optimizer.add_param_group({"params": model.parameters()})
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:586]: Why do we need rows 612 to 614?

File: tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py [hunk 7]
```diff
@@ -159,10 +159,16 @@ def build_toy_model_and_optimizer(model_type: str, init_model_with_meta_device:
...
             )
             fsdp_unit_modules = [te.pytorch.TransformerLayer]
-        toy_adam = Adam(params=toy_model.parameters(), lr=0.01)
 
     # Return the toy model, optimizer, and FSDP unit modules.
-    return toy_model, toy_adam, fsdp_unit_modules
+    return toy_model, fsdp_unit_modules
+
+
+def build_toy_optimizer(model: torch.nn.Module):
```
Comment (shjwudp) [inline tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py]: Can we remove `build_toy_optimizer` funciton? Initializing `Adam` inline seems more concise.

File: megatron/core/distributed/fsdp/src/README.md [hunk 8]
```diff
@@ -62,20 +62,34 @@ Transform your PyTorch model to use Fully Sharded Data Parallelism with just a f
 
 ```python
 import torch
-from megatron_fsdp import fully_shard
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/README.md:65]: Aligning with FSDP2: only the `fully_shard_model` and `fully_shard_optimizer` APIs should be demonstrated.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py (hunk 6 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:586]: FYI for those who have the same question, since we construct the model in fully_shard(), any optimizer that the user passes in hasn't been registered on the distributed parameters yet, so we must clear out the optimizer state.

File: megatron/core/distributed/fsdp/src/README.md (hunk 8 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/README.md:65]: Done, also updated the detailed API to use the `fully_shard_*` path, as well as added guidance on when to use the `fully_shard` all-in-one path.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 9]
```diff
@@ -57,10 +58,19 @@ class ShardingStrategy(IntEnum):
     OPTIM_GRADS_PARAMS = 3
 
 
+def experimental_api(func: Callable) -> Callable:
```
Comment (jaredcasper) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:61]: Why is this needed specifically here, isn't there a global one?

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:61]: Don't want to pip install MCore (`megatron.core.utils.experimental_api`) for MFSDP!

(I'll remove it in a subsequent PR when this API becomes the new baseline.)

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:61]: Updated the docstring to explain why:
```
def experimental_api(func: Callable) -> Callable:
    """
    Mark a function or class as experimental API in Megatron CI/CD.

    TODO(@cspades): Copied from megatron.core.utils to avoid depending on MCore
    for Megatron-FSDP. Should remove when the API is no longer experimental.
    """
```

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:61]: @jaredcasper Bump, all good here?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py (hunk 6 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:586]: Writing here, this is outdated, we can avoid this by just following the README:
```
optimizer = fully_shard_optimizer(
    torch.optim.Adam(model.parameters(), lr=1e-3)
)
```
This will initialize the optimizer directly on the Megatron-FSDP initialized parameters.

If the optimizer has other parameter groups, there may be bugs.

- Need to add MegatronFSDP model parameters as a new parameter group, or extend the existing parameter group.
- Need to modify the `fully_shard_optimizer` code to look for Megatron-FSDP parameters, instead of just taking the first Parameter.
  - In pipeline parallel, you may have multiple MegatronFSDP models! 
- Need to make sure multi-parameter groups are supported in conjunction with Megatron-FSDP's optimizer state and DCP metadata, maybe this breaks the DCP distributed state checkpointing hook.
